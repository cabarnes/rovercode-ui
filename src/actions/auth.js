// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const UPDATE_VALID_AUTH = 'UPDATE_VALID_AUTH';
export const USER_LOGOUT = 'USER_LOGOUT';

// action creators
export const updateValidAuth = (isValidAuth) => ({
  type: UPDATE_VALID_AUTH,
  payload: isValidAuth,
});

export const logout = () => ({
  type: USER_LOGOUT,
});

// helper functions
export const checkAuthError = (dispatch) => (error) => {
  if (error.response && error.response.status === 401) {
    // Authentication is no longer valid
    dispatch(updateValidAuth(false));
  } else {
    throw error;
  }
};

export const authHeader = (cookies) => ({
  headers: {
    Authorization: `JWT ${cookies.get('auth_jwt')}`,
  },
});
