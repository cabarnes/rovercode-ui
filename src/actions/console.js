// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const APPEND = 'APPEND';
export const CLEAR = 'CLEAR';

// action creators
export const append = (message) => ({
  type: APPEND,
  payload: message,
});

export const clear = () => ({
  type: CLEAR,
});
