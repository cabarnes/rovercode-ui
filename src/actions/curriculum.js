// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_COURSES = 'FETCH_COURSES';
export const FETCH_COURSES_PENDING = `${FETCH_COURSES}_PENDING`;
export const FETCH_COURSES_FULFILLED = `${FETCH_COURSES}_FULFILLED`;
export const FETCH_COURSES_REJECTED = `${FETCH_COURSES}_REJECTED`;

// action creators
export const fetchCourses = (xhrOptions) => ({
  type: FETCH_COURSES,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/courses/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});
