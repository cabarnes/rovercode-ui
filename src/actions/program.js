// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const CLEAR_PROGRAMS = 'CLEAR_PROGRAMS';
export const FETCH_PROGRAMS = 'FETCH_PROGRAMS';
export const FETCH_PROGRAMS_PENDING = `${FETCH_PROGRAMS}_PENDING`;
export const FETCH_PROGRAMS_FULFILLED = `${FETCH_PROGRAMS}_FULFILLED`;
export const FETCH_PROGRAMS_REJECTED = `${FETCH_PROGRAMS}_REJECTED`;
export const REMOVE_PROGRAM = 'REMOVE_PROGRAM';
export const REMOVE_PROGRAM_PENDING = `${REMOVE_PROGRAM}_PENDING`;
export const REMOVE_PROGRAM_FULFILLED = `${REMOVE_PROGRAM}_FULFILLED`;
export const REMOVE_PROGRAM_REJECTED = `${REMOVE_PROGRAM}_REJECTED`;

// action creators
export const clearPrograms = () => ({
  type: CLEAR_PROGRAMS,
});

export const fetchPrograms = (xhrOptions) => ({
  type: FETCH_PROGRAMS,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/block-diagrams/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const removeProgram = (id, xhrOptions) => ({
  type: REMOVE_PROGRAM,
  payload: axios.delete(`/api/v1/block-diagrams/${id}/`, xhrOptions)
    .then(({ data }) => (
      data
    )),
});
