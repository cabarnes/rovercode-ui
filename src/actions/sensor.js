// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const CHANGE_LEFT_SENSOR_STATE = 'CHANGE_LEFT_SENSOR_STATE';
export const CHANGE_RIGHT_SENSOR_STATE = 'CHANGE_RIGHT_SENSOR_STATE';
export const CHANGE_LIGHT_SENSOR_READINGS = 'CHANGE_LIGHT_SENSOR_READINGS';

// Sensor States
export const COVERED = 1;
export const NOT_COVERED = 2;

// action creators
export const changeLeftSensorState = (state) => ({
  type: CHANGE_LEFT_SENSOR_STATE,
  payload: state,
});

export const changeRightSensorState = (state) => ({
  type: CHANGE_RIGHT_SENSOR_STATE,
  payload: state,
});

export const changeLightSensorReadings = (leftReading, rightReading) => ({
  type: CHANGE_LIGHT_SENSOR_READINGS,
  payload: { leftReading, rightReading },
});
