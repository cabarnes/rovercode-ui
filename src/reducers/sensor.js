import {
  CHANGE_LEFT_SENSOR_STATE,
  CHANGE_RIGHT_SENSOR_STATE,
  CHANGE_LIGHT_SENSOR_READINGS,
  NOT_COVERED,
} from '../actions/sensor';

export default function sensor(
  state = {
    left: NOT_COVERED,
    right: NOT_COVERED,
    leftLightSensorReading: -1,
    rightLightSensorReading: -1,
  },
  action,
) {
  switch (action.type) {
    case CHANGE_LEFT_SENSOR_STATE:
      return {
        ...state,
        left: action.payload,
      };
    case CHANGE_RIGHT_SENSOR_STATE:
      return {
        ...state,
        right: action.payload,
      };
    case CHANGE_LIGHT_SENSOR_READINGS:
      return {
        ...state,
        leftLightSensorReading: action.payload.leftReading,
        rightLightSensorReading: action.payload.rightReading,
      };
    default:
      return state;
  }
}
