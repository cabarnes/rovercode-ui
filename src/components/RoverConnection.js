import React, { Component } from 'react';
import {
  Button,
  Typography,
  Popover,
} from '@material-ui/core';
import { grey } from '@material-ui/core/colors';
import { withStyles } from '@material-ui/core/styles';
import { FormattedMessage, injectIntl } from 'react-intl';
import { hot } from 'react-hot-loader';
import PropTypes from 'prop-types';

import { EXECUTION_STOP } from '@/actions/code';
import { COVERED, NOT_COVERED } from '@/actions/sensor';

import LogoIcon from './LogoIcon';

class RoverConnection extends Component {
  constructor(props) {
    super(props);

    this.protocolMap = {
      'light-sens': this.handleLightSensor,
      'line-sens': this.handleLineSensor,
      'dist-sens': this.handleDistanceSensor,
      'ub-temp-sens': this.handleUBitTempSensor,
      'ub-light-sens': this.handleUBitLightSensor,
      accel: this.handleAccelerationSensor,
      gyro: this.handleGyroscopeSensor,
      'compass-sens': this.handleCompassSensor,
      'mag-sens': this.handleMagneticForceSensor,
      'battery-sens': this.handleBatterySensor,
      'dewpoint-sens': this.handleDewPointSensor,
    };

    this.state = {
      unsupportedPopoverAnchorElement: null,
      supportedPlatform: navigator && navigator.bluetooth,
    };
  }

  handlePopoverOpen = (event) => {
    this.setState({
      unsupportedPopoverAnchorElement: event.target,
    });
  };

  handlePopoverClose = () => {
    this.setState({
      unsupportedPopoverAnchorElement: null,
    });
  };

  handleLightSensor = (params) => {
    const { changeLeftSensorState, changeRightSensorState, write } = this.props;

    const [left, right] = params.split(',');
    write(`Light Sensor - L:${left} R:${right}`);
    if (parseInt(left, 10) > 500) {
      changeLeftSensorState(COVERED);
    } else {
      changeLeftSensorState(NOT_COVERED);
    }
    if (parseInt(right, 10) > 500) {
      changeRightSensorState(COVERED);
    } else {
      changeRightSensorState(NOT_COVERED);
    }
  }

  handleLineSensor = (params) => {
    const { write } = this.props;

    const [left, right] = params.split(',');

    write(`Line Sensor - L:${left} R:${right}`);
  }

  handleDistanceSensor = (params) => {
    const { write } = this.props;

    write(`Distance Sensor - ${params} mm`);
  }

  handleUBitTempSensor = (params) => {
    const { write } = this.props;

    write(`uBit Temperature Sensor - ${params} C`);
  }

  handleUBitLightSensor = (params) => {
    const { write } = this.props;

    write(`uBit Light Sensor - ${params}`);
  }

  handleAccelerationSensor = (params) => {
    const { write } = this.props;

    const [x, y, z] = params.split(',');

    write(`Acceleration Sensor - X:${x} mG Y:${y} mG Z:${z} mG`);
  }

  handleGyroscopeSensor = (params) => {
    const { write } = this.props;

    const [pitch, roll] = params.split(',');

    write(`Gryoscope Sensor - Pitch:${pitch} degrees Roll:${roll} degrees`);
  }

  handleCompassSensor = (params) => {
    const { write } = this.props;

    write(`Compass Sensor - ${params} degrees`);
  }

  handleMagneticForceSensor = (params) => {
    const { write } = this.props;
    const [x, y, z] = params.split(',');

    write(`Magnetic Force Sensor - X:${x} uT Y:${y} uT Z:${z} uT`);
  }

  handleBatterySensor = (params) => {
    const { write } = this.props;

    write(`Battery Sensor - ${params} mV`);
  }

  handleDewPointSensor = (params) => {
    const { write } = this.props;

    write(`Dew Point Sensor - ${params} C`);
  }

  connect = () => {
    const { scanForRover, connectToRover } = this.props;

    return scanForRover().then((rover) => {
      rover.value.addEventListener('gattserverdisconnected', this.onDisconnected);
      return connectToRover(rover.value, this.onMessage);
    });
  }

  onMessage = (event) => {
    const { write } = this.props;

    const receivedData = [];
    for (let i = 0; i < event.target.value.byteLength; i++) {
      receivedData[i] = event.target.value.getUint8(i);
    }

    const receivedString = String.fromCharCode.apply(null, receivedData);

    const [command, params] = receivedString.split(':');
    try {
      this.protocolMap[command](params);
    } catch (e) {
      write('Unknown rover message received.');
    }
  }

  onDisconnected = () => {
    const { changeExecutionState, disconnectFromRover, rover } = this.props;

    changeExecutionState(EXECUTION_STOP);
    disconnectFromRover(rover);
  }

  render() {
    const { intl, rover } = this.props;
    const { unsupportedPopoverAnchorElement, supportedPlatform } = this.state;

    const popupText = intl.formatMessage({
      id: 'app.rover_connection.unsupported_platform',
      description: 'Popup text for unsupported platform',
      defaultMessage: 'This browser or device is not supported. '
        + 'Try using Google Chrome or Microsoft Edge '
        + 'on a PC, Macbook, or Android tablet. Also, '
        + 'make sure Bluetooth is enabled.',
    });

    const PopoverMessage = withStyles(() => ({
      root: {
        pointerEvents: 'none',
      },
    }))(Popover);

    const ConnectionButton = withStyles(() => ({
      root: {
        backgroundColor: 'white',
        '&:disabled': {
          backgroundColor: grey[400],
        },
      },
    }))(Button);

    const NoRoverText = withStyles(() => ({
      root: {
        color: grey[600],
      },
    }))(Typography);

    const ConnectText = withStyles(() => ({
      root: {
        textTransform: 'uppercase',
      },
    }))(Typography);

    const DisconnectText = withStyles(() => ({
      root: {
        textTransform: 'uppercase',
        color: grey[600],
      },
    }))(Typography);

    const PopoverMessageText = withStyles((theme) => ({
      root: {
        padding: theme.spacing(2),
        pointerEvents: 'none',
      },
    }))(Typography);

    const NavBarLogo = withStyles((theme) => {
      let color = grey[800];
      if (!supportedPlatform) { color = theme.palette.warning.dark; }
      if (rover) { color = theme.palette.primary.main; }
      return {
        root: { color },
      };
    })(LogoIcon);

    if (rover) {
      return (
        <ConnectionButton
          size="large"
          variant="contained"
          disableElevation
          startIcon={<NavBarLogo style={{ fontSize: 50 }} />}
          onClick={this.onDisconnected}
        >
          <div>
            <Typography variant="h6">
              {` ${rover.name.slice(15, 20)}`}
              <br />
            </Typography>
            <DisconnectText color="primary">
              <FormattedMessage
                id="app.rover_connection.disconnect"
                description="Button label to disconnect from the rover"
                defaultMessage="Disconnect"
              />
            </DisconnectText>
          </div>
        </ConnectionButton>
      );
    }

    const popoverOpen = Boolean(unsupportedPopoverAnchorElement);

    const button = (
      <ConnectionButton
        size="large"
        variant="contained"
        disableElevation
        onClick={this.connect}
        startIcon={(<NavBarLogo style={{ fontSize: 50 }} />)}
        disabled={!supportedPlatform}
      >
        <div>
          <NoRoverText variant="h6">
            <FormattedMessage
              id="app.rover_connection.no_rover"
              description="Label when no rover is connected"
              defaultMessage="No Rover"
            />
            <br />
          </NoRoverText>
          <ConnectText color={supportedPlatform ? 'primary' : 'initial'}>
            <FormattedMessage
              id="app.rover_connection.connect"
              description="Button label to connect to the rover"
              defaultMessage="Connect"
            />
          </ConnectText>
        </div>
      </ConnectionButton>
    );

    return supportedPlatform ? button : (
      <>
        <span
          aria-owns={popoverOpen ? 'mouse-over-popover' : undefined}
          aria-haspopup="true"
          onMouseEnter={this.handlePopoverOpen}
          onMouseLeave={this.handlePopoverClose}
        >
          {button}
        </span>
        <PopoverMessage
          id="mouse-over-popover"
          open={popoverOpen}
          anchorEl={unsupportedPopoverAnchorElement}
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left',
          }}
          transformOrigin={{
            vertical: 'top',
            horizontal: 'left',
          }}
          onClose={this.handlePopoverClose}
          disableRestoreFocus
        >
          <PopoverMessageText variant="h6">{popupText}</PopoverMessageText>
        </PopoverMessage>
      </>
    );
  }
}

RoverConnection.defaultProps = {
  rover: null,
};

RoverConnection.propTypes = {
  rover: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }),
  connectToRover: PropTypes.func.isRequired,
  disconnectFromRover: PropTypes.func.isRequired,
  scanForRover: PropTypes.func.isRequired,
  changeExecutionState: PropTypes.func.isRequired,
  changeLeftSensorState: PropTypes.func.isRequired,
  changeRightSensorState: PropTypes.func.isRequired,
  write: PropTypes.func.isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default hot(module)(injectIntl(RoverConnection));
