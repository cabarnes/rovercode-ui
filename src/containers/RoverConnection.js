import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';

import {
  scan as roverScan,
  connect as roverConnect,
  disconnect as roverDisconnect,
} from '@/actions/rover';
import {
  changeLeftSensorState as actionChangeLeftSensorState,
  changeRightSensorState as actionChangeRightSensorState,
} from '@/actions/sensor';
import RoverConnection from '@/components/RoverConnection';

const mapStateToProps = ({ rover }) => ({ ...rover });
const mapDispatchToProps = dispatch => ({
  changeLeftSensorState: state => dispatch(actionChangeLeftSensorState(state)),
  changeRightSensorState: state => dispatch(actionChangeRightSensorState(state)),
  connectToRover: (rover, onMessage) => dispatch(roverConnect(rover, onMessage)),
  disconnectFromRover: rover => dispatch(roverDisconnect(rover)),
  scanForRover: () => dispatch(roverScan()),
});

const RoverConnectionContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(RoverConnection);

export default hot(module)(RoverConnectionContainer);
