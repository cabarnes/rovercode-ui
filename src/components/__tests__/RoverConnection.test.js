import React from 'react';
import { Button } from 'semantic-ui-react';
import { FormattedMessage } from 'react-intl';
import { shallow } from 'enzyme';
import RoverConnection from '../RoverConnection';

import { COVERED, NOT_COVERED } from '@/actions/sensor';

let changeLeftSensorState;
let changeRightSensorState;
let connectToRover;
let disconnectFromRover;
let scanForRover;
let rover;
let wrapper;

const generateDataView = (array) => {
  const buffer = new ArrayBuffer(array.length);
  const dataView = new DataView(buffer);
  for (let i = 0; i < array.length; i++) {
    dataView.setUint8(i, array[i]);
  }

  return dataView;
};

describe('The RoverConnection component', () => {
  beforeEach(() => {
    rover = {
      name: 'BBC micro:bit [abcde]',
    };

    changeLeftSensorState = jest.fn();
    changeRightSensorState = jest.fn();
    connectToRover = jest.fn(() => Promise.resolve({}));
    disconnectFromRover = jest.fn();
    scanForRover = jest.fn(() => Promise.resolve({ value: rover }));

    wrapper = shallow(
      <RoverConnection
        changeLeftSensorState={changeLeftSensorState}
        changeRightSensorState={changeRightSensorState}
        connectToRover={connectToRover}
        disconnectFromRover={disconnectFromRover}
        scanForRover={scanForRover}
        rover={rover}
      />,
    );
  });

  test('renders on the page with no errors', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(FormattedMessage).prop('defaultMessage')).toEqual('Disconnect from');
    expect(wrapper.find(Button).children().at(1).text()).toBe(' abcde');
  });

  test('renders connect button when not connected', () => {
    wrapper = shallow(
      <RoverConnection
        changeLeftSensorState={changeLeftSensorState}
        changeRightSensorState={changeRightSensorState}
        connectToRover={connectToRover}
        disconnectFromRover={disconnectFromRover}
        scanForRover={scanForRover}
      />,
    );

    expect(wrapper.find(FormattedMessage).prop('defaultMessage')).toEqual('Connect to rover');
  });

  test('changes sensor state on message', () => {
    const leftSensor1 = [
      108, 101, 102, 116, 45, 115, 101, 110, 115, 111, 114, 58, 49,
    ];

    wrapper.instance().onMessage({
      target: {
        value: generateDataView(leftSensor1),
      },
    });

    expect(changeLeftSensorState).toHaveBeenCalledWith(COVERED);
    expect(changeRightSensorState).not.toHaveBeenCalled();

    changeRightSensorState.mockReset();
    changeLeftSensorState.mockReset();

    const leftSensor0 = [
      108, 101, 102, 116, 45, 115, 101, 110, 115, 111, 114, 58, 48,
    ];

    wrapper.instance().onMessage({
      target: {
        value: generateDataView(leftSensor0),
      },
    });

    expect(changeLeftSensorState).toHaveBeenCalledWith(NOT_COVERED);
    expect(changeRightSensorState).not.toHaveBeenCalled();

    changeRightSensorState.mockReset();
    changeLeftSensorState.mockReset();

    const rightSensor1 = [
      114, 105, 103, 104, 116, 45, 115, 101, 110, 115, 111, 114, 58, 49,
    ];

    wrapper.instance().onMessage({
      target: {
        value: generateDataView(rightSensor1),
      },
    });

    expect(changeRightSensorState).toHaveBeenCalledWith(COVERED);
    expect(changeLeftSensorState).not.toHaveBeenCalled();

    changeRightSensorState.mockReset();
    changeLeftSensorState.mockReset();

    const rightSensor0 = [
      114, 105, 103, 104, 116, 45, 115, 101, 110, 115, 111, 114, 58, 48,
    ];

    wrapper.instance().onMessage({
      target: {
        value: generateDataView(rightSensor0),
      },
    });

    expect(changeRightSensorState).toHaveBeenCalledWith(NOT_COVERED);
    expect(changeLeftSensorState).not.toHaveBeenCalled();

    changeRightSensorState.mockReset();
    changeLeftSensorState.mockReset();

    const invalid = [
      114, 105,
    ];

    wrapper.instance().onMessage({
      target: {
        value: generateDataView(invalid),
      },
    });

    expect(changeRightSensorState).not.toHaveBeenCalled();
    expect(changeLeftSensorState).not.toHaveBeenCalled();
  });

  test('connects to rover', async () => {
    await wrapper.instance().connect();

    expect(scanForRover).toHaveBeenCalled();
    expect(connectToRover).toHaveBeenCalledWith(rover, wrapper.instance().onMessage);
  });

  test('disconnects from rover', () => {
    wrapper.find(Button).simulate('click');

    expect(disconnectFromRover).toHaveBeenCalledWith(rover);
  });
});
