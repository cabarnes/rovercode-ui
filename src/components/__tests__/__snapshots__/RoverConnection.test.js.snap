// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The RoverConnection component renders on the page with no errors 1`] = `
<Fragment>
  <Button
    as="button"
    fluid={true}
    onClick={[Function]}
    primary={true}
  >
    <FormattedMessage
      defaultMessage="Disconnect from"
      id="app.rover_list.disconnect"
      values={Object {}}
    />
     abcde
  </Button>
</Fragment>
`;
